### PR TITLE
feat: leave subconversation when leaving MLS conference WPB-636

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-636" title="WPB-636" target="_blank">StoryWPB-636</a>  [iOS] Leave subgroup after leaving conference call
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
When leaving an MLS conference, we must also leave the subconversation hosting the conference. In this PR:

- add action to leave a subconversation
- add mls service method to leave the subconversation (calling the action and wiping the conversation)
- when leaving the call, invoke the mls service method to leave the subconversation
- to handle stale conferences (due to a crash during a call), also leave the subconversation when we detect an incoming call, terminated call, or a missed call.

#### Test Coverage

- unit tests asserting leave subconversation action
- unit tests asserting leaving subconversation
- unit tests asserting leaving subconversation if needed
- unit tests asserting WireCallCenter invokes methods to leave the subconversation

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift [hunk 1]
```diff
@@ -664,6 +664,40 @@ extension WireCallCenterV3 {
...
+            }
+
+            Task {
+                try await mlsService.leaveSubconversation(
+                    parentQualifiedID: QualifiedID(uuid: id.identifier, domain: domain),
+                    parentGroupID: parentGroupID,
+                    subconversationType: .conference
+                )
+            }
+        }
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: This implementation seems very similar to `leaveStaleConferenceIfNeeded` in `WireCallCenterV3+MLS`. I see that the main difference is that we don't call the same method from `mlsService` (`leaveSubconversation` vs `leaveSubconversationIfNeeded`). Could you help me understand the difference between the two and why one is used over the other?

File: wire-ios-data-model/Source/MLS/MLSService.swift [hunk 2]
```diff
@@ -1255,6 +1270,91 @@ public final class MLSService: MLSServiceInterface {
...
+                parentGroupID: parentGroupID,
+                subconversationType: subconversationType
+            )
+        }
+
+        if
+            let subConversationGroupID = subconverationGroupIDRepository.fetchSubconversationGroupID(
+                forType: subconversationType,
+                parentGroupID: parentGroupID
+            ),
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSService.swift:1292]: should the subconversation be removed from the repository after being left?

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: When we close a call we know we should leave the sub conversation, so we invoke `mlsService.leaveSubconversation` directly. This will leave the subconversation on the backend and then wipe the subgroup in CC. This is the happy path, where the self client is present when the call ends (by either side).

Suppose the self client crashes while in a call. When it relaunches, the call may be ongoing still, but the self client isn't part of the call anymore. It is however still part of the subconversation (backend) and subgroup (CC). To correct this state, we need to detect whether the subconversation has gone stale. This is done for instance by observing an incoming call state, which the self client will receive from the ongoing call. In this case we can detect that we're part of the subconversation but not the call, so we correct the state by leaving the subconversation.

`mlsService.leaveSubconversationIfNeeded` contains the logic for checking if we're part of a subconversation, and assuming that we shouldn't be, will leave the subconversation if we are. It does so by invoking `mlsService.leaveSubconversation`.

Perhaps the naming between the mls service methods and the call center methods is a bit confusing.

File: wire-ios-data-model/Source/MLS/MLSService.swift (hunk 2 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/MLS/MLSService.swift:1292]: yes, fixed

File: wire-ios-data-model/Tests/MLS/MLSServiceTests.swift [hunk 3]
```diff
@@ -1872,6 +1872,172 @@ class MLSServiceTests: ZMConversationTestsBase, MLSServiceDelegate {
         XCTAssertEqual(subconversationGroupID.parentGroupID, parentID)
     }
 
+    func test_LeaveSubconversation() async throws {
```
Comment (netbe) [inline wire-ios-data-model/Tests/MLS/MLSServiceTests.swift:1875]: this can be taken in another PR, but I think we should split up this tests class > 1800 lines

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift (hunk 1 above)
Comment (David-Henner) [reply wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: Yes perhaps renaming the methods or adding comments could help bring more clarity

File: wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/LeaveSubconversationActionHandler.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+final class LeaveSubconversationActionHandler: ActionHandler<LeaveSubconversationAction> {
+
+    // MARK: - Request
+
+    override func request(
+        for action: Action,
+        apiVersion: APIVersion
+    ) -> ZMTransportRequest? {
+        var action = action
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/LeaveSubconversationActionHandler.swift:30]: why do we need to redefine as var here ?

are the .fail methods mutating?

File: wire-ios-sync-engine/Tests/Source/Calling/WireCallCenterV3Tests.swift [hunk 5]
```diff
@@ -68,14 +68,16 @@ class WireCallCenterV3Tests: MessagingTest {
     override func setUp() {
         super.setUp()
 
+        BackendInfo.domain = "wire.com"
```
Comment (netbe) [inline wire-ios-sync-engine/Tests/Source/Calling/WireCallCenterV3Tests.swift:72]: we should avoid changing this directly/alone. change the storage before so it does not affect other tests
```
BackendInfo.storage = UserDefaults(suiteName: UUID().uuidString)!
```

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift [hunk 6]
```diff
@@ -664,6 +664,40 @@ extension WireCallCenterV3 {
...
+        guard
+            let context = uiMOC,
+            let syncContext = context.zm_sync,
+            let domain = id.domain ?? BackendInfo.domain,
+            let conversation = ZMConversation.fetch(
+                with: id.identifier,
+                domain: domain,
+                in: context
+            ),
+            conversation.messageProtocol == .mls,
```
Comment (netbe) [inline wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: should this be in the +MLS extension?

File: wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/LeaveSubconversationActionHandler.swift (hunk 4 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/LeaveSubconversationActionHandler.swift:30]: Correct, reporting the results are mutating.

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: I refactored the methods a bit and changed the names, also added comments.

File: wire-ios-sync-engine/Tests/Source/Calling/WireCallCenterV3Tests.swift (hunk 5 above)
Comment (johnxnguyen) [reply wire-ios-sync-engine/Tests/Source/Calling/WireCallCenterV3Tests.swift:72]: fixed

File: wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift (hunk 6 above)
Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift]: moved

**Review by netbe — APPROVED** [review]

**Review by David-Henner — APPROVED** [review]
